### PR TITLE
Handle Network Exceptions during app startup (AIC-480)

### PR DESCRIPTION
**Pull request by sam33rdhakal**
Updates:
* Fixes app crash during app data download (handles network exceptions)
* Refactors the splash logic of handling error

### CODE REVIEW

File: base/src/main/java/edu/artic/base/ObservableExtensions.kt [hunk 1]
```diff
@@ -41,4 +43,20 @@ fun <T> Observable<Result<T>>.mapWithDefault(onErrorValue : T) : Observable<T> {
...
             .map { (original, _) -> original }
-}
\ No newline at end of file
+}
+
+/**
+ * Extension method for handling network errors.
+ * TODO:: Use localized the error message.
+ */
+fun <T> Observable<T>.handleNetworkError(): Observable<T> {
```
Comment (Cliabhach) [inline base/src/main/java/edu/artic/base/ObservableExtensions.kt]: Can we use string resources for the `NetworkException` messages? Passing a context or resources in as parameter might work.

File: splash/src/main/kotlin/edu/artic/splash/SplashActivity.kt [hunk 2]
```diff
@@ -39,33 +43,39 @@ class SplashActivity : BaseViewModelActivity<SplashViewModel>(), TextureView.Sur
...
+                    if (BuildConfig.DEBUG) {
+                        percentText.visibility = View.VISIBLE
+                        percentText.text = errorMessage
+                    }
+
+                    /**
+                     * Display alert with error message.
+                     * TODO:: localized the error strings.
+                     */
+                    errorDialog?.dismiss()
```
[the remark this answers is not in the document]
Comment (sam33rdhakal) [reply splash/src/main/kotlin/edu/artic/splash/SplashActivity.kt]: This is tracked in the different ticket. So not the part of this ticket.

File: splash/src/main/kotlin/edu/artic/splash/SplashActivity.kt [hunk 3]
```diff
@@ -39,33 +43,39 @@ class SplashActivity : BaseViewModelActivity<SplashViewModel>(), TextureView.Sur
...
+                    errorDialog = AlertDialog.Builder(this, R.style.ErrorDialog)
+                            .setTitle(resources.getString(R.string.errorDialogTitle))
+                            .setMessage(errorMessage)
+                            .setPositiveButton(getString(android.R.string.ok)) { dialog, _ ->
+                                dialog.dismiss()
```
[the remark this answers is not in the document]
Comment (sam33rdhakal) [reply splash/src/main/kotlin/edu/artic/splash/SplashActivity.kt]: Again, this work will be done in another ticket.

File: base/src/main/java/edu/artic/base/NetworkException.kt [hunk 4]
```diff
@@ -0,0 +1,8 @@
+package edu.artic.base
+
+
+/**
+ * @author Sameer Dhakal (Fuzz)
+ */
+
+class NetworkException(override var message: String) : Exception(message)
```
[the remark this answers is not in the document]
Comment (sam33rdhakal) [reply base/src/main/java/edu/artic/base/NetworkException.kt]: Good idea. Updated.

File: base/src/main/java/edu/artic/base/ObservableExtensions.kt (hunk 1 above)
Comment (sam33rdhakal) [reply base/src/main/java/edu/artic/base/ObservableExtensions.kt]: Updated. Passes the context.

**Comment by Cliabhach** [comment]
@sam33rdhakal I cleaned up the commit history a little - let me know what you think.

**Pull request merged**